tests/gui/measure/test_hardwarebase.py

```diff
@@ -0,0 +1,156 @@
+"""Tests for module hardwarebase of viperleed.gui.measure."""
+
+__authors__ = (
+    'Michele Riva (@michele-riva)',
+    'Florian Dörr (@FlorianDoerr)',
+    )
+__copyright__ = 'Copyright (c) 2019-2025 ViPErLEED developers'
+__created__ = '2025-10-13'
+__license__ = 'GPLv3+'
+
+import abc
+import sys
+import types
+import pytest
+
+from viperleed.gui.measure.hardwarebase import class_from_name
+from viperleed.gui.measure.hardwarebase import get_devices
+from viperleed.gui.measure.hardwarebase import import_with_sub_modules
+
+
+@pytest.fixture
+def fake_pkg(mocker):
+    """
+    Create a fake package structure like:
+    fakepkg/
+        sub_a.py
+    and mock importlib.import_module and pkgutil.iter_modules so that
+    import_with_sub_modules and related functions can discover them.
+    """
+    base = 'fakepkg'
+    full_base = 'viperleed.gui.measure.fakepkg'
+
+    pkg = types.ModuleType(base)
+    pkg.__path__ = ['dummy_path']
+    pkg.__package__ = full_base
+
+    sys.modules[base] = pkg
+    sys.modules[full_base] = pkg
+
+    submodules = {}
+
+    def add_submodule(name, module):
+        sys.modules[f'{base}.{name}'] = module
+        sys.modules[f'{full_base}.{name}'] = module
+        submodules[name] = module
+
+    # Mock importlib.import_module.
+    def fake_import(name, package=None):
+        return sys.modules[name]
+
+    mocker.patch('importlib.import_module', side_effect=fake_import)
+
+    # Mock pkgutil.iter_modules to reflect whatever submodules exist.
+    def fake_iter_modules(path):
+        return [types.SimpleNamespace(name=n) for n in submodules.keys()]
+
+    mocker.patch('pkgutil.iter_modules', side_effect=fake_iter_modules)
+
+    pkg.add_submodule = add_submodule
+    pkg.submodules = submodules
+
+    # Define submodule containing a class A.
+    sub_a = types.ModuleType('fakepkg.sub_a')
+    class A:
+        pass
+    A.__module__ = sub_a.__name__
+    sub_a.A = A
+    pkg.add_submodule('sub_a', sub_a)
+
+    return pkg
+
+
+# Tests for import_with_sub_modules
+def test_import_with_sub_modules_yields_main_and_submodules(fake_pkg):
+    result = list(import_with_sub_modules('fakepkg'))
+    names = [m.__name__ for m in result]
+    assert 'fakepkg' in names
+    assert 'fakepkg.sub_a' in names
```

Comment:
It would be best to probably test equality rather than inclusion here:

```suggestion
    expect_names = [
        'fakepkg',        # First the top-level module
        'fakepkg.sub_a',  # Then all its submodules
        ]
    assert names == expect_names
```